crates/fiber-lib/src/store/pub_sub/store_with_pub_sub.rs

```diff
@@ -0,0 +1,286 @@
+use ckb_types::packed;
+use ractor::port::OutputPortSubscriber;
+
+use super::{InvoiceUpdatedPayload, PaymentUpdatedPayload, StorePublisher, StoreUpdatedEvent};
+#[cfg(feature = "watchtower")]
+use crate::watchtower::{WatchtowerStore, WatchtowerStoreDeref};
+use crate::{
+    fiber::{
+        channel::{ChannelActorStateStore, ChannelActorStateStoreDeref},
+        gossip::{GossipMessageStore, GossipMessageStoreDeref},
+        graph::NetworkGraphStateStore,
+        history::{Direction, TimedResult},
+        network::{NetworkActorStateStore, NetworkActorStateStoreDeref},
+        payment::{Attempt, AttemptStatus, PaymentSession, PaymentStatus},
+        types::Hash256,
+    },
+    invoice::{CkbInvoice, CkbInvoiceStatus, InvoiceError, InvoiceStore, PreimageStore},
+};
+
+#[derive(Clone, Debug)]
+pub struct StoreWithPubSub<S> {
+    pub(crate) inner: S,
+    publisher: StorePublisher,
+}
+
+impl<S> StoreWithPubSub<S> {
+    pub fn new(store: S) -> Self {
+        Self::new_with_publisher(store, StorePublisher::default())
+    }
+
+    pub fn new_with_publisher(store: S, publisher: StorePublisher) -> Self {
+        Self {
+            inner: store,
+            publisher,
+        }
+    }
+
+    pub(crate) fn publish(&self, event: StoreUpdatedEvent) {
+        self.publisher.publish(event);
+    }
+
+    pub fn subscribe(&self, subscriber: OutputPortSubscriber<StoreUpdatedEvent>) {
+        self.publisher.subscribe(subscriber);
+    }
+}
+
+impl<T: NetworkActorStateStore> NetworkActorStateStoreDeref for StoreWithPubSub<T> {
+    type Target = T;
+
+    fn network_actor_state_store_deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T: ChannelActorStateStore> ChannelActorStateStoreDeref for StoreWithPubSub<T> {
+    type Target = T;
+
+    fn channel_actor_state_store_deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<S> InvoiceStore for StoreWithPubSub<S>
+where
+    S: InvoiceStore,
+{
+    fn get_invoice(&self, id: &Hash256) -> Option<CkbInvoice> {
+        self.inner.get_invoice(id)
+    }
+
+    fn insert_invoice(
+        &self,
+        invoice: CkbInvoice,
+        preimage: Option<Hash256>,
+    ) -> Result<(), InvoiceError> {
+        let invoice_hash = *invoice.payment_hash();
+        self.inner.insert_invoice(invoice, preimage)?;
+
+        self.publish(StoreUpdatedEvent::new_invoice_updated_event(
+            invoice_hash,
+            InvoiceUpdatedPayload::Open,
+        ));
+        Ok(())
+    }
+
+    fn update_invoice_status(
+        &self,
+        id: &Hash256,
+        status: CkbInvoiceStatus,
+    ) -> Result<(), InvoiceError> {
+        let _span = tracing::info_span!("update_invoice_status", invoice_hash = ?id).entered();
+
+        self.inner.update_invoice_status(id, status)?;
+        let payload_opt = status.try_into().ok().or_else(|| match status {
```

Comment:
how about we add a field(something like `event_publisher`) in `Store`, and keep to use store here, so that when we need to notify some events, we just call `event_publisher.send_event` here.

seems the current implementation mixed the logic of sending events with current storing logic here.

and we also need to add delegating for all kinds of store, seems not good.

Reply:
Seems like a good idea. We can do transform in another layer of pub/sub.